File: 02-two_pointers/00011-Container_with_Most_Water/00011-container_with_most_water.py
# 11: Container with Most Water
# https://leetcode.com/problems/container-with-most-water/


class Solution:
    # SOLUTION
    def maxArea(self, height: list[int]) -> int:
        l = 0
        r = len(height) - 1
        result = 0

        while l<r:
            result = max(result, (r-l)*min(height[l], height[r]))
            if height[l]<height[r]: l+=1
            else: r-=1
        
        return result


if __name__ == "__main__":
    o = Solution()

    # INPUT
    height: list[int] = [1,8,6,2,5,4,8,3,7]

    # OUTPUT
    result = o.maxArea(height)
    print(result)

File: 02-two_pointers/00011-Container_with_Most_Water/00011-container_with_most_water.cpp
// 11: Container with Most Water
// https://leetcode.com/problems/container-with-most-water/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        int result = 0;

        while (l<r) {
            result = max(result, (r-l)*min(height[l], height[r]));
            if (height[l]<height[r]) l++;
            else r--;
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> height = {1,8,6,2,5,4,8,3,7};

    // OUTPUT
    auto result = o.maxArea(height);
    cout<<result<<endl;

    return 0;
}